File: 63. Unique Paths II.cpp
class Solution2 {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
    	int rows = obstacleGrid.size(); 
    	if(rows <= 0)	throw new runtime_error("invalid input");
    	int cols = obstacleGrid[0].size();
    	if(cols <= 0)	throw new runtime_error("invalid input");
    	if(obstacleGrid[0][0] == 1)	return 0;
    	vector<int> counts = vector<int>(cols, 1);
    	for(int i = 0; i < rows; ++i){
    		for(int j = 0; j < cols; ++j){
    			if(obstacleGrid[i][j] == 1)
    				counts[j] = 0;
    			else{
    				if(i == 0 && j > 0)
    					counts[j] = counts[j - 1] == 0 ? 0 : 1;
    				else if(j == 0 && i > 0)
    					counts[j] = counts[j] == 0 ? 0 : 1;
    				else
    					counts[j] = counts[j - 1] + counts[j];
    			}
    		}
    	}
    	return counts[cols - 1];
    }
};

File: 63. Unique Paths II.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        res = [[0] * n for i in range(m)]
        if obstacleGrid[0][0] != 1:
            res[0][0] = 1
        for i in range(1, m):
            res[i][0] = res[i-1][0]
            if obstacleGrid[i][0] == 1:
                res[i][0] = 0;
        for i in range(1, n):
            res[0][i] = res[0][i-1]
            if obstacleGrid[0][i] == 1:
                res[0][i] = 0;
        for i in range(1,m):
            for j in range(1,n):
                res[i][j] = res[i][j-1] + res[i-1][j]
                if obstacleGrid[i][j] == 1:
                    res[i][j] =0
        return res[m-1][n-1]
        
